### PR TITLE
Add new saved filters dock panel

### PR DESCRIPTION
**Pull request by cristianonicolai**

### CODE REVIEW

**Comment by cristianonicolai** [comment]
@nmirasch @bsremac please take a look. Just be aware that the duplicate code is quite intentional to leave the current UI as is until we can fully remove the current tabs.

**Comment by bsremac** [comment]
@cristianonicolai Am I the only one getting these errors when building _kie-wb-distributions/kie-wb-parent_ off this PR?
```
[INFO]    Tracing compile failure path for type 'org.drools.workbench.screens.guided.rule.client.editor.ActionValueEditor'
[INFO]       [ERROR] Errors in 'jar:file:/home/bsremac/.m2/repository/org/drools/drools-wb-guided-rule-editor-client/7.7.0-SNAPSHOT/drools-wb-guided-rule-editor-client-7.7.0-SNAPSHOT-sources.jar!/org/drools/workbench/screens/guided/rule/client/editor/ActionValueEditor.java'
[INFO]          [ERROR] Line 462: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 414: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 428: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]    Tracing compile failure path for type 'org.drools.workbench.screens.guided.rule.client.editor.ConstraintValueEditor'
[INFO]       [ERROR] Errors in 'jar:file:/home/bsremac/.m2/repository/org/drools/drools-wb-guided-rule-editor-client/7.7.0-SNAPSHOT/drools-wb-guided-rule-editor-client-7.7.0-SNAPSHOT-sources.jar!/org/drools/workbench/screens/guided/rule/client/editor/ConstraintValueEditor.java'
[INFO]          [ERROR] Line 663: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 676: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 611: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 644: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]          [ERROR] Line 597: The method addAttributeWithHelp(String, String, String, Button) is undefined for the type FormStylePopup
[INFO]    [ERROR] Aborting compile due to errors in some input files
```

**Comment by cristianonicolai** [comment]
@bsremac I've rebased https://github.com/kiegroup/appformer/pull/175, it should fix.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    public void init() {
+        loadSavedFilters();
+    }
+
+    protected void loadSavedFilters() {
+        preferencesService.call((MultiGridPreferencesStore multiGridPreferencesStore) -> {
+            if (multiGridPreferencesStore == null) {
+                multiGridPreferencesStore = new MultiGridPreferencesStore(getGridGlobalPreferencesKey());
+            }
+            this.multiGridPreferencesStore = multiGridPreferencesStore;
```
Comment (bsremac) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Just a suggestion:
```
 this.multiGridPreferencesStore = ofNullable(multiGridPreferencesStore).orElse(new MultiGridPreferencesStore(getGridGlobalPreferencesKey()));
```
Looks more concise to me. Although the statement becomes a bit lengthy for a one-liner. Maybe breaking it at the assignment point if you agree on the change?

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        tabSettingsValues.put(FILTER_TABLE_SETTINGS,
+                              dataSetEditorManager.getTableSettingsToStr(tableSettings));
+        tabSettingsValues.put(NewTabFilterPopup.FILTER_TAB_NAME_PARAM,
+                              tableSettings.getTableName());
+        tabSettingsValues.put(NewTabFilterPopup.FILTER_TAB_DESC_PARAM,
+                              tableSettings.getTableDescription());
+
+        saveNewTabSettings(tableSettings.getKey(),
+                           tabSettingsValues);
+
```
Comment (bsremac) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Just a suggestion, I would create a separate method which does the above and then call it in _initTabFilter_ before ```addNewTab(tableSettings)``` is called.
In that way, we would have the entire tab initialization grouped in one method, named accordingly _initTabFilter_ and let _addNewTab_ method do only a call to the view method which adds the saved filter.
Like this, it's a bit difficult to follow where the initialization stops and where we only use the values we initialized.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        } else {
+            existingGrids.stream().map(key -> new SavedFilter(key,
+                                                              getSavedFilterNameFromKey(key))).forEach(f -> view.addSavedFilter(f));
+        }
+    }
+
+    protected String getSavedFilterNameFromKey(final String key) {
+        final HashMap<String, Object> settings = multiGridPreferencesStore.getGridSettings(key);
+        if (settings == null) {
+            return null;
```
Comment (bsremac) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Would having ```null``` be acceptable case when creating a new instance of _SavedFilter_?

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+        }
+
+        return (String) settings.get(NewTabFilterPopup.FILTER_TAB_NAME_PARAM);
+    }
+
+    protected void removeSavedFilter(final String key) {
```
Comment (bsremac) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Comparing this method with the one below, the method's name doesn't seem quite right. The method below actually calls ```view.removeSavedFilter(savedFilter)``` which in turn makes the name logical.
On the other hand, this one here only does a preparation for filter removal, in the worst case. So I'm thinking if its possible to point out that difference with how we name them. 
Maybe we should just name this one ```removeFilter```, as search tab for example can't be qualified as _saved filter_.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    @Override
+    public String getGridGlobalPreferencesKey() {
+        return PROCESS_INSTANCES_WITH_VARIABLES_INCLUDED_LIST_PREFIX;
+    }
+
+    /*-------------------------------------------------*/
+    /*---              DashBuilder                   --*/
+    /*-------------------------------------------------*/
+    @Override
+    public FilterSettings createTableSettingsPrototype() {
```
Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java]: Some of the prototype methods are only one-liners, calling the basic settings method with ```null``` argument, whereas some have an independent body. Is there are a reason for it?

File: jbpm-wb-executor-service/jbpm-wb-executor-service-client/src/main/java/org/jbpm/workbench/es/client/editors/errorlist/ExecutionErrorListSavedFiltersPresenter.java [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                                    acknowledged ? "1" : "0"));
+        }
+
+        builder.filterOn(true,
+                         true,
+                         true);
+        builder.tableOrderEnabled(true);
+        builder.tableOrderDefault(COLUMN_ERROR_DATE,
+                                  DESCENDING);
+        addCommonColumnSettings(builder);
```
Comment (bsremac) [inline jbpm-wb-executor-service/jbpm-wb-executor-service-client/src/main/java/org/jbpm/workbench/es/client/editors/errorlist/ExecutionErrorListSavedFiltersPresenter.java]: All other presenters except this one lack a separate method for setting columns only. Maybe we should make it a common thing?

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        return createStatusSettings(ProcessInstance.STATE_COMPLETED);
+    }
+
+    public FilterSettings createAbortedTabSettings() {
+        return createStatusSettings(ProcessInstance.STATE_ABORTED);
+    }
+
+    @Override
+    public void initDefaultFilters() {
+        //Filter status Active
```
Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java:62]: These comments are present in some presenters and in some not

**Comment by bsremac** [comment]
@cristianonicolai So the new saved tabs are read-only at the moment, right? They can't be interacted with.

**Comment by cristianonicolai** [comment]
@bsremac correct, filters are "read only" for now until I get to remove the tabs, breaking these changes so reviewing is at least more manageable.

**Comment by bsremac** [comment]
+1

**Review by nmirasch — APPROVED** [review]
For me is a very good starting point to remove tab filters. I understand that is the first step and for sure it takes a lot of the current tab code regarding the filters creation. I let here my approval, and try to be aware about the @bsremac  comments about the code.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java (hunk 1 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Done

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java (hunk 2 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Did some refactor here, hope its better now :)

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java (hunk 3 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Definitely not, throwing exception now.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java (hunk 4 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFiltersPresenter.java]: Renamed method _removeSavedFilterFromPreferences_

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java (hunk 5 above)
Comment (cristianonicolai) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java]: Refactored the filter settings creation, its a bit more inline now.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java (hunk 7 above)
Comment (cristianonicolai) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/editors/instance/list/ProcessInstanceListSavedFiltersPresenter.java:62]: Added comments to all places.

**Comment by cristianonicolai** [comment]
@bsremac @nmirasch please take another look here, did a bit of refactoring, breaking some of the "cloned" code to a more definitive version.

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFilter.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+package org.jbpm.workbench.common.client.filters;
+
+import org.jboss.errai.databinding.client.api.Bindable;
+
+@Bindable
+public class SavedFilter {
+
+    private String name;
+
+    private String key;
```
Comment (bsremac) [inline jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFilter.java]: Would just change the order of these two fields to match the order they are initialized in the constructor and the order getters and setters are defined.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/perspectives/ProcessDefinitionsPerspective.java [hunk 9]
```diff
@@ -44,4 +44,9 @@ public String getPerspectiveId() {
     public String getBasicFiltersScreenId() {
         return null;
     }
+
+    @Override
+    public String getSavedFiltersScreenId() {
+        return null;
```
Comment (bsremac) [inline jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/perspectives/ProcessDefinitionsPerspective.java:50]: I'm, in general, against returning ```null```, but maybe this is a good case for an exception to that rule.
I can't think of any other approach except for introducing ```Optional``` as a return value, but that might just complicate the code and make it harder to read.

**Review by bsremac — APPROVED** [review]

**Comment by bsremac** [comment]
@cristianonicolai You just have to resolve these conflicts first.

File: jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/perspectives/ProcessDefinitionsPerspective.java (hunk 9 above)
Comment (cristianonicolai) [reply jbpm-wb-process-runtime/jbpm-wb-process-runtime-client/src/main/java/org/jbpm/workbench/pr/client/perspectives/ProcessDefinitionsPerspective.java:50]: +1

File: jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFilter.java (hunk 8 above)
Comment (cristianonicolai) [reply jbpm-wb-common/jbpm-wb-common-client/src/main/java/org/jbpm/workbench/common/client/filters/SavedFilter.java]: Done

**Comment by cristianonicolai** [comment]
@bsremac should be good for merge now.

**Review by bsremac — APPROVED** [review]

**Pull request merged**